test/groupBy.test.ts

```diff
@@ -0,0 +1,16 @@
+import { groupBy } from '../es';
+
+const byGrade = groupBy((student: { score: number; name: string }) => {
+  const score = student.score;
+  return score < 65 ? 'F' : score < 70 ? 'D' : score < 80 ? 'C' : score < 90 ? 'B' : 'A';
+});
+const students = [
+  { name: 'Abby', score: 84 },
+  { name: 'Eddy', score: 58 },
+  { name: 'Jack', score: 69 }
+];
+
+const grouped = byGrade(students);
+(grouped.C ?? []).length;
+// @ts-expect-error
+grouped.C.length;
```

Comment:
The goal of the test are to tests the types, and not the actual function. So what is here does nothing, the way `tsd` works basically just ignores this file as is. We can use this as a base though. Add some test cases like this:

```typescript
import { expectType } from 'tsd';
import { groupBy } from '../es';

const byGrade = groupBy((student: { score: number; name: string }) => {
  const score = student.score;
  return score < 65 ? 'F' : score < 70 ? 'D' : score < 80 ? 'C' : score < 90 ? 'B' : 'A';
});

expectType<(list: readonly { score: number; name: string }[]) => Partial<Record<'A' | 'B' | 'C' | 'D' | 'E' | 'F', { score: number; name: string }[]>>>(byGrade);
expectType<Partial<Record<'A' | 'B' | 'C' | 'D' | 'E' | 'F', { score: number; name: string }[]>>>(byGrade([]));
```

Check out some of the other existing test files for inspiration on how to cover all the possible cases

Reply:
It does work, actually! If you e.g. remove `Partial` from the return type, then `grouped.C.length` is allowed and thus the `@ts-expect-error` would throw a typecheck error. This is how DefinitelyTyped `@types/ramda` was tested.
In fact I'd argue it is a more resilient way of writing the tests because it focuses on what code is allowed and what code isn't, instead of expecting a long specific type that pretty much just repeats what the type definition says.

I can of course change it to use `expectType` assertions to be consistent with the other tests. But just a word of warning, this kind of tests will go red on pretty much any change to the types, and not just when a change was made that actually breaks something for the consumer of the types.

Reply:
Let me know if you really want me to add some `expectType`s. My testing heart will be in slight pain extending the tests to be deliberately brittle, but I understand that codebase consistency also matters :D

Reply:
> It does work, actually! If you e.g. remove Partial from the return type, then grouped.C.length is allowed and thus the @ts-expect-error would throw a typecheck error

Oh yeah didn't think about how that would throw a typecheck error. Cool

Reply:
Just convert what you have to use `tsd` and we'll be good

Reply:
added some tsd type assertions